File: python/src/theoryforge/embedding.py
"""Opt-in embedding-based construct-redundancy screen (API_SPEC.md Part E).

PARITY-EXEMPT: results depend on a user-supplied embedding function whose outputs are not
deterministic across model versions or SDKs. This is the assistive counterpart to the
deterministic lexical screen in `redundancy.py`; it is excluded from the parity contract and CI.
"""
from __future__ import annotations

import math
from collections.abc import Callable, Sequence

from . import _resources
from ._num import rnd


def _list(d: dict, key: str) -> list:
    v = d.get(key)
    return v if isinstance(v, list) else []


def _cosine(a: Sequence[float], b: Sequence[float]) -> float:
    num = sum(x * y for x, y in zip(a, b, strict=False))
    da = math.sqrt(sum(x * x for x in a))
    db = math.sqrt(sum(y * y for y in b))
    if da == 0 or db == 0:
        return 0.0
    return num / (da * db)


def embedding_redundancy(T, embedder: Callable[[str], Sequence[float]],
                         threshold: float | None = None) -> list[dict]:
    """Pairwise cosine similarity of embedded construct definitions.

    `embedder` maps a definition string to a numeric vector. Returns one record per unordered
    construct pair, sorted by descending similarity then (a, b), with a `review`/`ok` flag.
    """
    T = T.data if hasattr(T, "data") else T
    if threshold is None:
        threshold = _resources.checklist()["thresholds"]["redundancy_similarity_max"]
    cons = _list(T, "constructs")
    vecs = [(c.get("id", ""), embedder(c.get("definition", ""))) for c in cons]
    rows = []
    for i in range(len(vecs)):
        for j in range(i + 1, len(vecs)):
            sim = rnd(_cosine(vecs[i][1], vecs[j][1]), 6)
            rows.append({
                "a": vecs[i][0], "b": vecs[j][0],
                "cosine": sim, "flag": "review" if sim >= threshold else "ok",
            })
    rows.sort(key=lambda r: (-r["cosine"], r["a"], r["b"]))
    return rows

File: r/theoryforge/R/embedding.R
#' Opt-in embedding-based construct-redundancy screen.
#'
#' This screen is parity-exempt because its results depend on a user-supplied
#' embedding function whose outputs are not deterministic across model versions
#' or SDKs. It is the assistive counterpart to the deterministic lexical screen
#' in \code{tf_redundancy_check}, and it is excluded from the parity contract
#' and CI.
#' See API_SPEC.md section 23.
#' @name embedding
#' @keywords internal
NULL

# Cosine similarity of two numeric vectors; 0 when either has zero norm.
.tf_cosine <- function(a, b) {
  a <- as.numeric(a)
  b <- as.numeric(b)
  num <- sum(a * b)
  da <- sqrt(sum(a * a))
  db <- sqrt(sum(b * b))
  if (da == 0 || db == 0) {
    return(0.0)
  }
  num / (da * db)
}

#' Embedding-based pairwise construct-redundancy screen
#'
#' For every unordered pair of constructs, embeds each definition with the
#' supplied \code{embedder} and computes the cosine similarity (rounded to 6
#' decimals). Returns a data frame sorted by descending cosine then
#' \code{(a, b)}, flagging pairs at or above \code{threshold} for review. This
#' assistive screen complements the deterministic lexical
#' [tf_redundancy_check()] and is parity-exempt. See API_SPEC.md section 23.
#'
#' @param theory A theory object (named list), e.g. from [tf_read()].
#' @param embedder A function mapping a definition string to a numeric vector.
#' @param threshold Cosine threshold for the \code{"review"} flag; defaults to
#'   the checklist's \code{redundancy_similarity_max}.
#' @return A data frame with columns \code{a}, \code{b}, \code{cosine},
#'   \code{flag}.
#' @examples
#' theory <- tf_theory("demo-1", "A demonstration theory") |>
#'   tf_add_construct("c_arousal", "Arousal", "bodily activation") |>
#'   tf_add_construct("c_threat", "Threat", "appraised danger")
#' # A toy deterministic embedder: bag-of-words counts over a fixed vocabulary.
#' vocab <- c("bodily", "activation", "appraised", "danger")
#' embedder <- function(def) {
#'   words <- strsplit(tolower(def), "\\s+")[[1]]
#'   vapply(vocab, function(w) sum(words == w), numeric(1))
#' }
#' tf_embedding_redundancy(theory, embedder)
#' @export
tf_embedding_redundancy <- function(theory, embedder, threshold = NULL) {
  T <- theory
  if (is.null(threshold)) {
    threshold <- tf_checklist()$thresholds$redundancy_similarity_max
  }
  cons <- .tf_list(T, "constructs")
  ids <- vapply(cons, function(c) .tf_str(c, "id"), character(1))
  vecs <- lapply(cons, function(c) embedder(.tf_str(c, "definition")))

  a <- character(0)
  b <- character(0)
  cosine <- numeric(0)
  n <- length(cons)
  if (n >= 2L) {
    for (i in seq_len(n - 1L)) {
      for (j in (i + 1L):n) {
        a <- c(a, ids[[i]])
        b <- c(b, ids[[j]])
        cosine <- c(cosine, .tf_rnd(.tf_cosine(vecs[[i]], vecs[[j]]), 6))
      }
    }
  }
  flag <- ifelse(cosine >= threshold, "review", "ok")
  df <- data.frame(a = a, b = b, cosine = cosine, flag = flag,
                   stringsAsFactors = FALSE)
  if (nrow(df) > 0L) {
    ord <- order(-df$cosine, df$a, df$b, method = "radix")
    df <- df[ord, , drop = FALSE]
    rownames(df) <- NULL
  }
  df
}
